ValueTuple: adding Deconstruct and other tuple interop methods for System.Tuple

jcouv opened the pull request:
This PR is meant for early feedback.
It only includes extension methods for ref tuples (`System.Tuple`) of length 21. The three extension methods are Deconstruct, ToRefTuple and ToValueTuple.
I will add the deconstruct and helper methods for tuples from 1 to 20 before asking to merge.

@stephentoub @KevinRansom @VSadov for feedback.

jcouv commented:
Added implementation for tuples of length 1 through 20. This is now ready for review. Thanks.

jcouv commented:
test Innerloop Windows_NT Release Build and Test please

stephentoub commented on src/System.ValueTuple/src/System/ValueTuple/TupleExtensions.cs:
```diff
@@ -0,0 +1,906 @@
+            this Tuple<T1> value,
+            out T1 item1)
+        {
+            item1 = value.Item1;
```
In all of these cases, do we want/need to check ```value``` against null and throw an ArgumentNullException, or are we ok with a NullReferenceException here?

jcouv replied:
I think NRE is good. Deconstruct is not meant to be called by the user, but it will be called by compiler generated code. So having an ArgumentNullException that reports the argument called 'value' won't help the user much to track down the problem.

stephentoub replied:
Seems reasonable.

stephentoub commented on src/System.ValueTuple/src/System/ValueTuple/TupleExtensions.cs:
```diff
@@ -0,0 +1,906 @@
+        /// <summary>
+        /// Deconstruct a propertly nested <see cref="Tuple"/> with 1 elements.
+        /// </summary>
+        public static void Deconstruct<T1>(
```
Do we expect these methods to be called directly by code?  I'm wondering if we should use attributes to hide them or something.  No strong opinion.

jcouv replied:
That's a good idea. Let me see what attributes to put.

stephentoub commented on src/System.ValueTuple/src/System/ValueTuple/TupleExtensions.cs:
```diff
@@ -0,0 +1,906 @@
+        /// </summary>
+        public static void Deconstruct<T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12, T13, T14, T15, T16, T17, T18, T19, T20, T21>(
+            this Tuple<T1, T2, T3, T4, T5, T6, T7, Tuple<T8, T9, T10, T11, T12, T13, T14, Tuple<T15, T16, T17, T18, T19, T20, T21>>> value,
+            out T1 item1, out T2 item2, out T3 item3, out T4 item4, out T5 item5, out T6 item6, out T7 item7, out T8 item8, out T9 item9, out T10 item10, out T11 item11, out T12 item12, out T13 item13, out T14 item14, out T15 item15, out T16 item16, out T17 item17, out T18 item18, out T19 item19, out T20 item20, out T21 item21)
```
How did we land on 21 being the upper bound of what we provide Deconstruct overloads for?

jcouv replied:
This is the result of discussion with F# team, who we used to represent the System.Tuple users. The longest tuples they found in any project was 9. So we initially thought 1 nesting (14 elements total), but that didn't feel like much buffer. So we went to 2 nestings.

stephentoub replied:
Has there been any discussions about allowing the compiler to generate the equivalent of these methods rather than having them in a library (specifically for ValueTuple/Tuple, as I realize it's a general pattern that can be implemented by other types)?

jcouv replied:
Yes. The C# compiler does have knowledge of ValueTuple and we discussed having special knowledge of System.Tuple as well. But we decided against it because it is relatively more expensive and not such a common scenario. 
So we decided to rely on the normal deconstruction mechanism that types other than ValueTuple have to rely on.

VSadov replied:
Note that "generate the equivalent of these methods" would not work well. If there are methods, then one can bind them to delegate, so where the methods would live? It would need to be a common location per program (not per assembly).

The other viable alternative is to make compiler intrinsically familiar with System.Tuple and not require Deconstruct methods in these scenarios per language spec. That is a half way to supporting System.Tuple as a secondary tuple compatible type and seems too heavy.

stephentoub replied:
> The other viable alternative is to make compiler intrinsically familiar with System.Tuple and not require Deconstruct methods in these scenarios per language spec

FWIW, this is actually what I meant.

stephentoub commented on src/System.ValueTuple/src/System/ValueTuple/TupleExtensions.cs:
```diff
@@ -0,0 +1,906 @@
+        /// Make a properly nested <see cref="Tuple"/> from a properly nested <see cref="ValueTuple"/> with 1 element.
+        /// </summary>
+        public static Tuple<T1>
+            ToRefTuple<T1>(
```
Should this just be "ToTuple" rather than "ToRefTuple"?

stephentoub replied:
Also, what's the scenario for having these To* methods?

jcouv replied:
Yes, I think I'll change to `ToTuple` instead. I got other feedback that `ToRefTuple` is ugly and also doesn't line up with the type names.

As far as usage, our goal is to allow conversions between ref and value tuples. But conversion operators can't do the job above 8, because a conversion operator needs to be implemented in the type that it produces or the type that it comes from. But `VT<T1, ..., VT<T8>>` is not declared anywhere (so there is nowhere to put the operators).
Instead, the fallback is to have the user explicitly convert.

stephentoub commented on src/System.ValueTuple/tests/ValueTuple/ExtensionsTests.cs:
```diff
@@ -0,0 +1,701 @@
+    [Fact]
+    public void ConvertToRef1()
+    {
+        var refTuple = Tuple.Create(-1);
```
Why construct a new tuple if we're just going to overwrite it a few lines down?

jcouv replied:
That's an easy way to make sure the type of refTuple is correct.
I could change to `Tuple<int, ...> refTuple` instead, but this was more convenient.

stephentoub replied:
Ok.  The latter is what I had in mind, but for tests it's not a big deal.

jcouv commented:
Renamed `ToRefTuple` to `ToTuple` and added `EditorBrowsable(Never)` attribute on `Deconstruct`.
I'm ambivalent about `DebuggerNonUserCode` or `DebuggerHidden`  or `DebuggerStepThrough`. Let me know if you think they are useful here.

VSadov commented:
Since the methods are going to be in an external to the user library, I do not think DebuggerXX attributes are relevant.

jcouv commented:
test Innerloop OSX Release Build and Test please

jcouv commented:
@stephentoub This is ready to merge if no other questions or feedback.

stephentoub commented on src/System.ValueTuple/src/System/ValueTuple/TupleExtensions.cs:
```diff
@@ -0,0 +1,929 @@
+        }
+
+        /// <summary>
+        /// Deconstruct a propertly nested <see cref="Tuple"/> with 2 elements.
```
Nit: search and replace propertly=>properly

jcouv replied:
Yes

stephentoub commented on src/System.ValueTuple/src/System/ValueTuple/TupleExtensions.cs:
```diff
@@ -0,0 +1,929 @@
+            item5 = value.Item5;
+            item6 = value.Item6;
+            item7 = value.Item7;
+            item8 = value.Rest.Item1;
```
I know I asked about the NRE case above for value itself.  But what about Rest (and Rest.Rest)?  Or is it not possible for Rest to be null?

jcouv replied:
The ctor for Tuple'8 has a check, and I think NRE would be fine in this case anyways.

```C#
   if (!(rest is ITuple)) {
                throw new ArgumentException(Environment.GetResourceString("ArgumentException_TupleLastArgumentNotATuple"));
            }
```

stephentoub commented:
LGTM.  Though it'd be good for some other reviews before we merge it.

jcouv commented:
@stephentoub Sounds good. Who should I ping for additional reviews?

stephentoub commented:
> Who should I ping for additional reviews?

@VSadov? @AlekseyTs?  Just another set of eyes from folks familiar with the plans around tuples.

VSadov commented:
LGTM

jcouv commented:
test Innerloop Windows_NT Debug Build and Test please
test Innerloop OSX Debug Build and Test please
